### PR TITLE
chore(asm): bases of api security in appsec

### PR DESCRIPTION
**Pull request by christophe-papazian**
Add the bases needed by API security in appsec.
No new feature in this PR, as the new functions are not used in any supported framework yet.

- improve asm_request_context to handle response headers and bodies
- add a the possibility of global callbacks (not request related)
- callbacks are now called with the content of the asm context values to get easy access to them from anywhere
- add route and response.body in `SPAN_DATA_NAMES ` constants
- add api_security directory with main code (already deployed and tested for the api_security demo)
- add some tests for the type schema computation

## Checklist

- [x] Change(s) are motivated and described in the PR description.
- [x] Testing strategy is described if automated tests are not included in the PR.
- [x] Risk is outlined (performance impact, potential for breakage, maintainability, etc).
- [x] Change is maintainable (easy to change, telemetry, documentation).
- [x] [Library release note guidelines](https://ddtrace.readthedocs.io/en/stable/releasenotes.html) are followed. If no release note is required, add label `changelog/no-changelog`.
- [x] Documentation is included (in-code, generated user docs, [public corp docs](https://github.com/DataDog/documentation/)).
- [x] Backport labels are set (if [applicable](../docs/contributing.rst#release-branch-maintenance))

## Reviewer Checklist

- [x] Title is accurate.
- [x] No unnecessary changes are introduced.
- [x] Description motivates each change.
- [x] Avoids breaking [API](https://ddtrace.readthedocs.io/en/stable/versioning.html#interfaces) changes unless absolutely necessary.
- [x] Testing strategy adequately addresses listed risk(s).
- [x] Change is maintainable (easy to change, telemetry, documentation).
- [x] Release note makes sense to a user of the library.
- [x] Reviewer has explicitly acknowledged and discussed the performance implications of this PR as reported in the benchmarks PR comment.
- [x] Backport labels are set in a manner that is consistent with the [release branch maintenance policy](../docs/contributing.rst#release-branch-maintenance)

### CODE REVIEW

File: ddtrace/appsec/_api_security/api_manager.py [hunk 1]
```diff
@@ -0,0 +1,173 @@
...
+    from typing import Dict
+    from typing import Optional
+
+
+log = get_logger(__name__)
+metrics = Metrics(namespace="datadog.api_security")
+_sentinel = object()
+
+
+class TooLarge(Exception):
```
[the remark this answers is not in the document]
Comment (christophe-papazian) [reply ddtrace/appsec/_api_security/api_manager.py]: Good idea. Done.

File: tests/appsec/api_security/test_schema_fuzz.py [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+        ({"toast": "touch", "tomato": "tommy"}, [{"toast": [8]}, {"truncated": True}]),
+        ({"foo": {"bar": 42}}, [{"foo": [{"bar": [0]}]}]),
+    ],
+)
+def test_limits(obj, res):
+    assert equal_without_meta(build_schema(obj, max_depth=2, max_girth=1, max_types_in_array=1), res)
+
+
+if __name__ == "__main__":
+    import atheris
```
Comment (avara1986) [inline tests/appsec/api_security/test_schema_fuzz.py:75]: What is that package? typo?

Comment (christophe-papazian) [reply tests/appsec/api_security/test_schema_fuzz.py]: Good catch. This was for local testing. Removed.

File: ddtrace/appsec/_api_security/api_manager.py [hunk 3]
```diff
@@ -0,0 +1,173 @@
...
+        add_context_callback(self._schema_callback, global_callback=True)
+
+    def _should_collect_schema(self, env):
+        if self.INTERVAL_PER_ROUTE <= 0.0:
+            return True
+        method = env.waf_addresses.get(SPAN_DATA_NAMES.REQUEST_METHOD)
+        route = env.waf_addresses.get(SPAN_DATA_NAMES.REQUEST_ROUTE)
+        # Framework is not fully supported
+        if not method or not route:
+            log.debug("unsupported groupkey for api security [method %s] [route %s]", bool(method), bool(route))
```
Comment (nizox) [inline ddtrace/appsec/_api_security/api_manager.py:108]: this log is on the request path, it's probably better to use `self._log_limiter`

File: ddtrace/appsec/_api_security/api_manager.py [hunk 4]
```diff
@@ -0,0 +1,173 @@
...
+
+        try:
+            if not self._should_collect_schema(env) or self._global_rate_limiter.limit() is RateLimitExceeded:
+                return
+        except Exception:
+            self._log_limiter.limit(log.warning, "Failed to sample request for schema generation", exc_info=True)
+
+        # we need the request content type on the span
+        try:
+            headers = env.waf_addresses.get(SPAN_DATA_NAMES.REQUEST_HEADERS_NO_COOKIES, _sentinel)
```
Comment (nizox) [inline ddtrace/appsec/_api_security/api_manager.py:142]: shall we move this to the appsec processor?

**Review by nizox — DISMISSED** [review]

File: ddtrace/appsec/_api_security/api_manager.py (hunk 4 above)
Comment (christophe-papazian) [reply ddtrace/appsec/_api_security/api_manager.py:142]: probably a good idea. I'll talk about that with the asm python team.

Comment (christophe-papazian) [reply ddtrace/appsec/_api_security/api_manager.py:142]: After some thoughts, we can do it later. For now, I would prefer that this new private feature doesn't change the regular behaviour of the tracer at all if not activated.

File: ddtrace/appsec/_asm_request_context.py [hunk 5]
```diff
@@ -142,6 +149,59 @@ def set_value(category, address, value):  # type: (str, str, Any) -> None
         asm_context_attr[address] = value
 
 
+def set_headers_response(headers):  # type: (Any) -> None
```
[the remark this answers is not in the document]
Comment (christophe-papazian) [reply ddtrace/appsec/_asm_request_context.py:152]: done

**Review by avara1986 — APPROVED** [review]

File: ddtrace/appsec/_api_security/api_manager.py [hunk 6]
```diff
@@ -0,0 +1,173 @@
...
+
+        self._schema_meter = metrics.get_meter("schema")
+        self._log_limiter = RateLimiter(limit_rate=1.0, raise_on_exceed=False)
+        self._global_rate_limiter = RateLimiter(
+            limit_rate=self.GLOBAL_RATE_LIMIT,
+            raise_on_exceed=False,
+        )
+        self._rate_limiter_by_route = dict()  # type: Dict[str, RateLimiter]
+
+        log.info("%s initialized", self.__class__.__name__)
```
Comment (avara1986) [inline ddtrace/appsec/_api_security/api_manager.py:91]: info? why not debug? ;)

File: ddtrace/appsec/_api_security/api_manager.py [hunk 7]
```diff
@@ -0,0 +1,173 @@
...
+        # type: () -> None
+        if cls._instance is None:
+            log.info("%s not enabled", cls.__name__)
+            return
+
+        log.info("Disabling %s", cls.__name__)
+        cls._instance.stop()
+        cls._instance = None
+        metrics.disable()
+        log.info("%s disabled", cls.__name__)
```
Comment (avara1986) [inline ddtrace/appsec/_api_security/api_manager.py:73]: info? why not debug? ;)

File: ddtrace/appsec/_api_security/api_manager.py [hunk 8]
```diff
@@ -0,0 +1,173 @@
...
+
+    INTERVAL_PER_ROUTE = 15.0  # seconds between requests
+
+    _instance = None  # type: Optional[APIManager]
+
+    @classmethod
+    def enable(cls):
+        # type: () -> None
+        if cls._instance is not None:
+            log.info("%s already enabled", cls.__name__)
```
Comment (avara1986) [inline ddtrace/appsec/_api_security/api_manager.py:53]: info? why not debug? ;)

File: ddtrace/appsec/_api_security/api_manager.py [hunk 9]
```diff
@@ -0,0 +1,173 @@
...
+            return
+
+        log.info("Enabling %s", cls.__name__)
+        metrics.enable()
+        cls._instance = cls()
+        cls._instance.start()
+        log.info("%s enabled", cls.__name__)
```
Comment (avara1986) [inline ddtrace/appsec/_api_security/api_manager.py:60]: ditto

File: ddtrace/appsec/_api_security/api_manager.py [hunk 10]
```diff
@@ -0,0 +1,173 @@
...
+        return r.limit() is not RateLimitExceeded
+
+    def _schema_callback(self, env):
+        if env.span is None:
+            return
+        root = env.span._local_root or env.span
+        if not root or any(meta_name in root._meta for _, meta_name, _ in self.COLLECTED):
+            return
+
+        root._metrics["_dd.appsec.api_security.enabled"] = 1.0
```
Comment (avara1986) [inline ddtrace/appsec/_api_security/api_manager.py:132]: can we move "_dd.appsec.api_security.enabled" to constants?

File: ddtrace/appsec/_asm_request_context.py [hunk 11]
```diff
@@ -142,6 +149,20 @@ def set_value(category, address, value):  # type: (str, str, Any) -> None
...
+    if headers is not None:
+        set_waf_address(SPAN_DATA_NAMES.RESPONSE_HEADERS_NO_COOKIES, headers, _ASM.get().span)
+
+
+def set_body_response(body_response):
+    from ddtrace.appsec.utils import parse_response_body
```
Comment (avara1986) [inline ddtrace/appsec/_asm_request_context.py:158]: why is it a closure this import? why not global?

File: ddtrace/appsec/utils.py [hunk 12]
```diff
@@ -182,3 +182,52 @@ def parse_message(msg):
...
+    )
+    if not content_type:
+        return
+
+    def access_body(bd):
+        if isinstance(bd, list) and isinstance(bd[0], (str, bytes)):
+            bd = bd[0][:0].join(bd)
+        if getattr(bd, "decode", False):
+            bd = bd.decode("UTF-8", errors="ignore")
+        if len(bd) >= 0x1000000:
```
Comment (avara1986) [inline ddtrace/appsec/utils.py:217]: why 16MB? 16MB depends of the SO?

Comment (christophe-papazian) [reply ddtrace/appsec/utils.py:217]: We had to choose a limit, Nicolas proposed that value. I should give it a name.

File: ddtrace/appsec/_api_security/api_manager.py (hunk 7 above)
Comment (christophe-papazian) [reply ddtrace/appsec/_api_security/api_manager.py:73]: you are right.

File: ddtrace/appsec/_api_security/api_manager.py [hunk 13]
```diff
@@ -0,0 +1,173 @@
...
+    _instance = None  # type: Optional[APIManager]
+
+    @classmethod
+    def enable(cls):
+        # type: () -> None
+        if cls._instance is not None:
+            log.info("%s already enabled", cls.__name__)
+            return
+
+        log.info("Enabling %s", cls.__name__)
```
[the remark this answers is not in the document]
Comment (christophe-papazian) [reply ddtrace/appsec/_api_security/api_manager.py:56]: It gives more info about where something could have gone wrong.

File: ddtrace/appsec/_api_security/api_manager.py (hunk 10 above)
Comment (christophe-papazian) [reply ddtrace/appsec/_api_security/api_manager.py:132]: yes !

**Review by avara1986 — APPROVED** [review]

**Pull request merged**